### ISSUE TITLE
@SpringBootTest requires Jackson in classpath (I have only GSON)

### ISSUE THREAD

**Issue by malafeev** [spring-projects/spring-boot#8168]
Spring boot version: 1.5.1.RELEASE

My Spring Boot based application uses GSON and I excluded jackson from dependencies:

```
<dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-web</artifactId>
            <exclusions>
                <exclusion>
                    <artifactId>jackson-databind</artifactId>
                    <groupId>com.fasterxml.jackson.core</groupId>
                </exclusion>
            </exclusions>
</dependency>

<dependency>
            <groupId>com.google.code.gson</groupId>
            <artifactId>gson</artifactId>
            <version>2.7</version>
</dependency>
```

It works fine.
But I cannot test it using @SpringBootTest.
I get error:

```
Caused by: java.lang.NoClassDefFoundError: com/fasterxml/jackson/core/JsonProcessingException
	at org.springframework.data.web.config.SpringDataWebConfiguration.extendMessageConverters(SpringDataWebConfiguration.java:123)
	at org.springframework.web.servlet.config.annotation.WebMvcConfigurerComposite.extendMessageConverters(WebMvcConfigurerComposite.java:143)
	at org.springframework.web.servlet.config.annotation.DelegatingWebMvcConfiguration.extendMessageConverters(DelegatingWebMvcConfiguration.java:122)
	at org.springframework.web.servlet.config.annotation.WebMvcConfigurationSupport.getMessageConverters(WebMvcConfigurationSupport.java:718)
	at org.springframework.web.servlet.config.annotation.WebMvcConfigurationSupport.requestMappingHandlerAdapter(WebMvcConfigurationSupport.java:525)
	at org.springframework.boot.autoconfigure.web.WebMvcAutoConfiguration$EnableWebMvcConfiguration.requestMappingHandlerAdapter(WebMvcAutoConfiguration.java:372)
```

Could you add support GSON for @SpringBootTest?

**Comment by wilkinsona** [comment]
> It works fine

As far as I can tell, that's not the case. If I exclude Jackson from `spring-boot-sample-data-rest` the application doesn't start _and_ its tests fail. This applies to both 1.4.x and 1.5.x.

Can you please provide a small sample that illustrates it working fine when you run your application but failing when you run its tests?

**Comment by wilkinsona** [comment]
I should also have said that Spring Data REST requires Jackson, i.e. `spring-data-rest-web` has a non-optional dependency on Jackson. The best that we will be able to do here is to make the auto-configuration of Spring Data REST back off when Jackson isn't on the classpath.

**Comment by malafeev** [comment]
I will provide sample. 
But from documentation there is support for Jackson and Gson. So I can choose.
Also look at https://github.com/spring-projects/spring-boot/issues/7431

**Comment by malafeev** [comment]
My project is not Spring Data REST.
It's plain Spring Boot. Only REST controllers. No data :)

**Comment by wilkinsona** [comment]
That's not what the stack trace says: 

```
org.springframework.data.web.config.SpringDataWebConfiguration.extendMessageConverters(SpringDataWebConfiguration.java:123)
```

`SpringDataWebConfiguration` is a class from Spring Data REST.

Please provide a sample.

**Comment by malafeev** [comment]
Sample:
https://github.com/malafeev/spring-boot-test


I found that problem in 

```
 <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-jpa</artifactId>
</dependency>
```

But I need this dependency for database access (using JPA).

**Comment by wilkinsona** [comment]
Thanks for the sample. Sorry, I was mistaken earlier. `SpringDataWebConfiguration` is part of Spring Data Commons.

The requirement for Jackson is a result of the [changes](https://github.com/spring-projects/spring-data-commons/commit/eec2b43fd08885ae4f3ebe87a5f86ef249970d8f) made for [DATACMNS-885](https://jira.spring.io/browse/DATACMNS-885). This change was included in Spring Data Ingalls which Boot 1.5 uses (1.4 uses Spring Data Hopper). I think this is a regression so I've opened [DATACMNS-987](https://jira.spring.io/browse/DATACMNS-987).

You see different behaviour in your tests versus at runtime because the classpaths are different. Specifically, `spring-boot-starter-test` depends on `com.jayway.jsonpath:json-path` which triggers the faulty behaviour in `SpringDataWebConfiguration`. You can work around the problem by adding an exclusion:

```xml
<dependency>
    <groupId>org.springframework.boot</groupId>
    <artifactId>spring-boot-starter-test</artifactId>
    <scope>test</scope>
    <exclusions>
        <exclusion>
            <artifactId>json-path</artifactId>
            <groupId>com.jayway.jsonpath</groupId>
        </exclusion>
    </exclusions>
</dependency>
```

**Comment by malafeev** [comment]
Thanks for workaround for this particular case.
But I need jsonpath in classpath for testing using MockMvc

```
this.mvc.perform(get("/v1/accounts/1").andExpect(jsonPath("$.name", is("test name")));
```

I hope [DATACMNS-987](https://jira.spring.io/browse/DATACMNS-987) will fix the issue.

**Comment by olivergierke** [comment]
That's fixed in Spring Data Commons now. Until that's released (probably just in time for Boot 1.5.1) adding Jackson in at least test scope should help working around that issue.

**Comment by malafeev** [comment]
Thank you!
